src/v/pandaproxy/schema_registry/avro.cc

```diff
@@ -379,6 +379,7 @@ ss::future<collected_schema> collect_schema(
               std::move(ss.schema));
         }
     }
+    // NOLINTNEXTLINE(bugprone-use-after-move)
```

Comment:
The safety argument for these is that the implementation of `canonical_schema` is such that it doesn't clobber `def()` on the original, right?

This feels like very much worth fixing, so that it's not a trap for any change to the implementation of the class -- I think it's worth a ticket to help us not forget, and perhaps target it for during 23.1.x stabilization before release.

Reply:
> The safety argument for these is that the implementation of canonical_schema is such that it doesn't clobber def() on the original, right?

I would describe as `std::move(schema).def()` doesn't clobber other fields in the original.

Say you have:

```
struct S {
   std::string a;
   std::string b;
};

void sink(S s) {
   something(std::move(s.a));
   something(std::move(s.b));
}
```

Clearly here we are cherry-picking stuff out of `s` to avoid copies. We own `s` so this is safe. What is happening PP/SR is that S is defined with an r-value qualified accessors:

```
class S {
   std::string a;
   std::string b;
public:
   auto take_a() && { return std::move(a); }
   auto take_b() && { return std::move(b); }
};
```

The goal here (please correct me if I'm wrong @BenPope) is to force call sites to experss that instances of `S` that are being cherry-picked from via `take_{a,b}` are end-of-life:

```
S s;
s.take_a() // doesn't compile
std::move(s).take_a() // works fine
```

> This feels like very much worth fixing, so that it's not a trap for any change to the implementation of the class -- I think it's worth a ticket to help us not forget, and perhaps target it for during 23.1.x stabilization before release.

Agree that we should fix it, but not sure i'd call it a trap since it requires the same analysis looking at the code. But I'm strongly in favor of it because `clang-tidy` can't reason about it and throws a false positive.

Reply:
Follow up ticket https://github.com/redpanda-data/redpanda/issues/8216